elephant/joint_isi_dithering_class.py

```diff
@@ -0,0 +1,462 @@
+'''
+Module to generate surrogates of spike trains by using the joint-ISI dithering.
+Features are provided to separate the preprocessing from the main process.
+
+Original implementation by: Peter Bouss [p.bouss@fz-juelich.de]
+:copyright: Copyright 2019 by the Elephant team, see `doc/authors.rst`.
+:license: Modified BSD, see LICENSE.txt for details.
+'''
+
+
+import numpy as np
+import quantities as pq
+import neo
+
+import elephant.spike_train_surrogates as surr
+import elephant.statistics as stats
+from scipy.ndimage import gaussian_filter
+
+try:
+    import matplotlib.pyplot as plt
+    HAVE_PLT = True
+except ImportError:
+    HAVE_PLT = False
+
+class Joint_ISI_Space:
+    '''
+    The class :class:`Joint_ISI_Space` is implemented for Joint-ISI dithering
+    as a continuation of the ideas of Louis et al. (2010) and Gerstein (2004).
+
+    When creating an class instance all necessary preprocessing steps are done,
+    to use the method dithering().
+
+    To make the dithering-procedure stable, the spiketrain needs to pass two
+    thresholds, that the Joint-ISI dithering is applied, if not uniform
+    dithering is used.
+
+    Attributes
+    ----------
+    st: neo.SpikeTrain
+        For this spiketrain the surrogates will be created
+
+    n_surr: int, optional
+        Number of surrogates to be created.
+        Default: 1
+    dither: pq.Quantity
+        The range of the dithering for the uniform dithering,
+        which is also used for the method 'window'.
+        Default: 15.*pq.ms
+    unit: pq.unit
+        The unit of the spiketrain in the output.
+        Default: pq.s
+    window_length: pq.Quantity
+        The extent in which the joint-ISI-distribution is calculated.
+        Default: 120*pq.ms
+    num_bins: int
+        The size of the joint-ISI-ditribution will be num_bins*num_bins.
```

Comment:
typo -> distribution
ps: why would you want to change this parameter?

Reply:
Because it gives me a finer grid on which the cumulative distribution functions are calculated. And thus the effects of the discretization are less. I think this is normally more import for the low ISI-regime. - Instead it could introduce a binsize parameter a just assert, that it is a multiple of the window length.